File: 359. Logger Rate Limiter.py
class Logger:
    '''
    This problem we should consider the space that we use to store the messages,
    so we should keep what is inputed in the last 10 seconds,
    and we should delete all previous messages
    '''
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.printed = set()
        self.queue = collections.deque()
        

    def shouldPrintMessage(self, timestamp: int, message: str) -> bool:
        """
        Returns true if the message should be printed in the given timestamp, otherwise returns false.
        If this method returns false, the message will not be printed.
        The timestamp is in seconds granularity.
        """
        while self.queue and timestamp - self.queue[-1][0] >= 10:
            poped_t, poped_m = self.queue.pop()
            self.printed.remove(poped_m)
            
        if message in self.printed:
            return False
        else:
            self.queue.appendleft((timestamp, message))
            self.printed.add(message)
            return True

File: 359. Logger Rate Limiter.java
class Logger {
    class Node {
        int time;
        String message;
        Node(int time, String message) {
            this.time = time;
            this.message = message;
        }
    }
    Deque<Node> deque;

    /** Initialize your data structure here. */
    public Logger() {
        deque = new ArrayDeque<>();
    }
    
    /** Returns true if the message should be printed in the given timestamp, otherwise returns false.
        If this method returns false, the message will not be printed.
        The timestamp is in seconds granularity. */
    public boolean shouldPrintMessage(int timestamp, String message) {
        while (!deque.isEmpty() && deque.peekFirst().time <= timestamp-10) {
            deque.pollFirst();
        }
        for (Node n : deque) {
            if (n.message.equals(message)) {
                return false;
            }
        }
        deque.offer(new Node(timestamp, message));
        return true;
    }
}

/**
 * Your Logger object will be instantiated and called as such:
 * Logger obj = new Logger();
 * boolean param_1 = obj.shouldPrintMessage(timestamp,message);
 */